pkg/kv/kvserver/batcheval/cmd_delete_range.go

```diff
@@ -123,112 +109,158 @@ func DeleteRange(
 			return result.Result{}, errors.AssertionFailedf(
 				"ReturnKeys can't be used with range tombstones")
 		}
+	}
+
+	hasPredicate := args.Predicates != (kvpb.DeleteRangePredicates{})
+	if hasPredicate && args.IdempotentTombstone {
+		return result.Result{}, errors.AssertionFailedf("IdempotentTombstone not compatible with Predicates")
+	}
 
-		desc := cArgs.EvalCtx.Desc()
-
-		maybeUpdateGCHint := func(res *result.Result) error {
-			if !args.UpdateRangeDeleteGCHint {
-				return nil
-			}
-			sl := MakeStateLoader(cArgs.EvalCtx)
-			hint, err := sl.LoadGCHint(ctx, readWriter)
-			if err != nil {
-				return err
-			}
-
-			updated := hint.ScheduleGCFor(h.Timestamp)
-			// If the range tombstone covers the whole Range key span, update the
-			// corresponding timestamp in GCHint to enable ClearRange optimization.
-			if args.Key.Equal(desc.StartKey.AsRawKey()) && args.EndKey.Equal(desc.EndKey.AsRawKey()) {
-				// NB: don't swap the order, we want to call the method unconditionally.
-				updated = hint.ForwardLatestRangeDeleteTimestamp(h.Timestamp) || updated
-			}
-			if !updated {
-				return nil
-			}
-
-			if err := sl.SetGCHint(ctx, readWriter, cArgs.Stats, hint); err != nil {
-				return err
-			}
-			res.Replicated.State = &kvserverpb.ReplicaState{
-				GCHint: hint,
-			}
+	switch {
+	case args.UseRangeTombstone && !hasPredicate:
+		return deleteRangeIdempotent(ctx, readWriter, cArgs)
+	case hasPredicate:
+		return deleteRangeWithPredicate(ctx, readWriter, cArgs, reply)
+	default:
+		return deleteRangeTransactional(ctx, readWriter, cArgs, reply)
+	}
+}
+
+func deleteRangeIdempotent(
```

Comment:
I'm not sure on the name of this one. Seems to me from the logic that we'll come here even if IdempotentTombstone is false?  I'm not sure if I have a better name. Perhaps `deleteRangeUsingTombostone` to line up to the underlying storage call. That too is a bit of a misnomer since deleteRangeWithPredicate also uses tombstones.

Reply:
Yeah, I agree this name is mediocre given the current assertions.

There are a bunch of tests that use range tombstones without setting the idempotent flag, but all of the production users I could find always set the idempotent flag. 

I wanted to keep this PR to refactoring only. But I could stack a second commit/PR that updates the tests and requires the idempotent flag to enter this code path.

Reply:
The name `deleteRangeUsingTombstone` also makes sense given the fact I intend to rework delete predicate so it does not require range tombstones.